src/userdev/java/net/minecraftforge/gradle/userdev/MinecraftUserRepo.java

```diff
@@ -973,7 +1072,42 @@ private File compileJava(File source, File... extraDeps) {
             compile.setDestinationDir(output);
             compile.setSource(source.isDirectory() ? project.fileTree(source) : project.zipTree(source));
 
-            compile.execute();
+            // What follows is a horrible hack to allow us to call JavaCompile
+            // from our dependency resolver.
+            // As described in https://github.com/MinecraftForge/ForgeGradle/issues/550,
+            // invoking Gradle tasks in the normal way can lead to deadlocks
+            // when done from a dependency resolver.
+
+            // To avoid these issues, we invoke the 'compile' method on JavaCompile
+            // using reflection.
+
+            // Normally, the output history is set by Gradle. Since we're bypassing
+            // the normal gradle task infrastructure, we need to do it ourselves.
+            compile.getOutputs().setHistory(new TaskExecutionHistory() {
+
+                @Override
+                public Set<File> getOutputFiles() {
+                    // We explicitly clear the output directory
+                    // ourselves, so it's okay that this is totally wrong.
+                    return Sets.newHashSet();
+                }
+
+                @Nullable
+                @Override
+                public OverlappingOutputs getOverlappingOutputs() {
+                    return null;
+                }
+
+                @Nullable
+                @Override
+                public OriginTaskExecutionMetadata getOriginExecutionMetadata() {
+                    return null;
+                }
+            });
+
+            Method compileMethod = JavaCompile.class.getDeclaredMethod("compile");
+            compileMethod.setAccessible(true);
+            compileMethod.invoke(compile);
```

Comment:
Considering the compile function is protected, we could make a subclass of JavaCompile and hard link it so that we get a compile error instead of runtime error if Gradle changes?